### ISSUE TITLE
proposal: Helm dynamic aliases

### ISSUE THREAD

**Issue by kfox1111** [helm/helm#3126]
TL:DR; Essentially it is a way for an end user to instantiate multiple versions of a subchart and give it the values for each instances.

I would like to have a new flag on dependencies in requirements.yaml which is a boolean such as:
dynamic-alias: true

When the dependency is set as a dynamic alias, the following behaviour is changed:
the values are consulted for the typical value section for the child chart. (either the aliased name if specified, or the charts name if alias is not specified).
1. If the section is blank, the child chart is not used at all. (treated like it is conditionally disabled)
2. If the section has data, the root of the section is a hash. The key is used to name the child chart instance. One instance per key. The values in the subsection is used for that instance of the object.
3. the subcharts release would be named after the release and the key. such as "parent-child".

As extra credit, :), perhaps there should be a 'defaults' key that is used to set defaults that effect all instances in the section. so like:

requirements.yaml:
```yaml
dependencies:
- name: mariadb
  version: 0.7.0
  repository: https://kubernetes-charts.storage.googleapis.com/
  dynamic-alias: true
```

values.yaml:
```
some_stuff:...
mariadb:
    defaults:
      pvc.create: true
    foo:
      memsize: 16g
    bar:
      memsize: 8g
```

Which would launch two mariadb instances with different memory sizes.

I have some charts that I currently launch as half a dozen instances of the chart, that I could use this functionality, and a tweak to the values file to treat as a single chart I upgrade when I want to add more instances that are 99% the same config as the rest. (prometheus node exporters with different node selectors/plugins per host type, flannel instances with different nic configs, etc)

**Comment by thomastaylor312** [comment]
@kfox1111 This sounds like an interesting idea, I worry about conflating `values.yaml` keys with keys that define the name of the chart. Also, I know you can use aliases to reuse a chart multiple times, so is this more to simplify the process for users?

**Comment by kfox1111** [comment]
how about something more env like?

mariadb:
  - name: defaults
    values: 
      pvc:
        create: true
  - name: foo
    values:
      memsize: 16g
  - name: bar
    values:
      memsize: 8g

I view requirements.yaml as more of a chart developer level feature. the users's values.yaml is for end users. using the existing aliases feature would require the user to download the package, extract it, tweak requirements and optionally rebuild the package. Plus would have to have a custom one per cluster potentially if each had different numbers of subcharts.

I think this feature would make it much easier to have more shareable charts.

**Comment by thomastaylor312** [comment]
That might not be a bad idea. This would definitely have to be something that would land in Helm 3 though

**Comment by omkensey** [comment]
Relatedly, it would be nice sometimes to be able to "push down" values from the top-level chart to the depended/child charts.  Sort of the reverse of child/parent import:

`parent-values.yaml:`
<pre>
foo: 7
bar: 12
<b>push:
  - to: child-a, child-b.solarsystems.quux
    moons: yes
    comets: no</b>
</pre>

Then both child-a and child-b would have the relevant values inserted in their contexts, like so (highlighted in `diff` style):

`child-a:`
```diff
  planets: yes
+ moons: yes
+ comets: no
```

`child-b`:
```diff
  solarsystems:
    fred:
      planets: no
      moons: no
      comets: yes
    quux:
      planets: yes
+     moons: yes
+     comets: no
```

**Comment by kfox1111** [comment]
@omkensey yeah, I really want something similar. I do aggregation charts all the time and it kind of sucks for the user to have to set the same value all over the place. but maybe we need one level of abstraction higher. rather then specify the values right in the push rule, that should just specify a section of the main values file to use. maybe like:

foo: 7
bar: 12
moons: yes
comets: no

#maybe this belongs in requires.yaml instead?)
push: #(or copy or defaults)
  - to: [child-a, child-b.solarsystems.quux]
    values:
      - name: moons #(if value is not specified, moons is used too)
      - name: foocomets
        value: comets #pushes the value of comets from the parent into the value of foocomets in the children

in a way, this isn't pushing values. its rearanging values inside the parents values.yaml? It copies the value to the child's subsection of the values file if there isn't already a value there specified by the user. This still allows the user to explicitly override....

maybe a more concrete example:
replicas: 3

with:
push:
  - to: [ws1, ws2]
    values:
      - name: replicas

the user can just set replica's and each child gets the same values. but the user can still do
--set replilcas=1 --set ws2.replicas=2

**Comment by omkensey** [comment]
I like the modification @kfox1111 came up with.

**Comment by kfox1111** [comment]
Now that helm's pushing ahead without crd release objects, templating out the release like was proposed during the summit isn't an option. I think we should go ahead and implement this feature then.

**Comment by bacongobbler** [comment]
I haven't seen any movement on this proposal since it was first opened 3 years ago. Closing as inactive.
